### ISSUE TITLE
Add alternative stellar yield tables

### ISSUE THREAD

**Issue by abensonca** [galacticusorg/galacticus#500]
Given that there are no strong constraints on the best yield tables to use it would make sense to have a few different options for these so that we can explore the variations resulting from modeling uncertainties.

This is relatively simple to do as we would just need to create different files containing the yields etc. for different models.

**Comment by abensonca** [comment]
From James Johnson:

> For massive stars, Limongi & Chieffi (2018) looked at yields at log10(Z / Zsun) = -3, -2, -1, and 0 at initial rotational velocities of v_rot = 0, 150, and 300 km/s. If you're looking for yields from the literature that go down to [Fe/H] ~ -3, I'd start there. The one caution with Limongi & Chieffi (2018) is that they have a simplified prescription for which massive stars explode (everything above 25 Msun collapses to a black hole). They've also got a relatively sparse mass grid. For those reasons I'd also look into Sukhbold et al.'s (2016) yields, which have a much more realistic prescription for stellar explodability and a much denser grid of ZAMS masses, but there you're limited to solar metallicity and non-rotating progenitors.

> For SN Ia yields, the papers I'm most familiar with are Iwamoto et al. (1999), Seitenzahl et al. (2013), and the 2021/2022 series from Sabrina Gronow. Classically, people assume that SN Ia yields don't depend on the metallicity of the progenitor population. The Gronow et al. series would argue the contrary -- that the changes in WD composition with metallicity affect the structure enough to impact yields, though as far as I can tell, they had to fix the Fe yields in their models to do the calculations, so it's hard to do much quantitatively with their results. Last fall, I wrote a paper with Chris Kochanek and Kris Stanek at OSU arguing that the metallicity dependence of the close binary fraction should produce more Ia's at low Z (with a ~Z^(-1/2) dependence), and that the implied rates as a function of mass line up with what's seen in, e.g., ASAS-SN and DES.

**Comment by abensonca** [comment]
Also, for the `stellarAstrophysicsFile` class we should process `XInclude` elements in the XML file so that it would be simple to combine different tables of yields etc. as needed. It then might make sense to take the different datasets currently compiled into the [static/stellarAstrophysics/Stellar_Properties_Compilation.xml](https://github.com/galacticusorg/datasets/blob/master/static/stellarAstrophysics/Stellar_Properties_Compilation.xml) file and separate them so that they can be combined as needed.

**Comment by abensonca** [comment]
Also from James Johnson:

> I'd also advocate for implementing a separate mode of chemical enrichment in which the user specifies (for each enrichment channel, e.g., CCSNe, SNe Ia) a yield for each element, which may vary with metallicity Z of the stellar pops, and a delay-time distribution. A lot of abundance modelers, myself included, are starting to move beyond tables from stellar models as the only option for parameterizing yields. That's primarily driven by the low number of stellar nucleosynthesis models available in the literature, and the models that do exist are quite rife with uncertainties to boot.

**Comment by abensonca** [comment]
> Also, for the `stellarAstrophysicsFile` class we should process `XInclude` elements in the XML file so that it would be simple to combine different tables of yields etc. as needed. It then might make sense to take the different datasets currently compiled into the [static/stellarAstrophysics/Stellar_Properties_Compilation.xml](https://github.com/galacticusorg/datasets/blob/master/static/stellarAstrophysics/Stellar_Properties_Compilation.xml) file and separate them so that they can be combined as needed.

This is actually already implemented - the `XML_Parse()` function handles this for us. So, we'd just need to split this file into components and replace is with one that `XIncludes` all of those components.

**Comment by abensonca** [comment]
From James Johnson no Ia delay times:

> Our most direct measurements come from comparisons of the cosmic star formation history with volumetric event rates as a function of redshift - [Maoz & Mannucci (2012)](https://ui.adsabs.harvard.edu/abs/2012PASA...29..447M) is a highly cited article doing that. Even then, they get measurements in only three bins of delay-time with sizable errors. My understanding is that the cosmic SFH is the more poorly constrained of the two, but both carry substantial uncertainties.

**Comment by abensonca** [comment]
The stellar properties file is [now](https://github.com/galacticusorg/datasets/commit/559f0fe14da04df71c9cac5b161b30971354f9d3) split into individual files, seperated by source (i.e. Heger & Woosley 2002 in one file, Marigo et al. 2001 in another, etc.). The standard compilation of stellar properties is now simply a file that uses `XInclude` to combine these separate files. This allows for more flexibility in combining different sets of stellar property models as needed.

**Comment by abensonca** [comment]
## Feasibility survey: alternative stellar yield tables

I went through James' suggestions and checked, for each paper, whether machine-readable tables actually exist and whether they map onto our current infrastructure. Short answer: **Limongi & Chieffi (2018) is an almost perfect fit and needs no new code at all**; Sukhbold et al. (2016) is a good complement but has a metallicity problem; the SN Ia suggestions are easy *except* where they carry a metallicity dependence, which our current format cannot express.

### What we have today

`stellarAstrophysicsFile` reads an XML file of `<star>` elements keyed on `initialMass` + `metallicity`, with optional `lifetime`, `ejectedMass`, `metalYieldMass` and `elementYieldMass<X>`, and interpolates irregularly in the (mass, metallicity) plane. Two things matter for what follows:

- The tabulated yields are **net** yields (the existing Portinari and Marigo files contain negative entries), and `ejectedMass` is the **gross** mass returned.
- `fileName` is already an input parameter, and `XML_Parse()` already honours `XInclude` (the standard compilation is now just a wrapper that includes the four per-source files). **So a new yield set is purely a datasets-repo change** — no Fortran required.

Current coverage, for reference:

| Source | Masses (M☉) | Z | Provides |
|---|---|---|---|
| Portinari, Chiosi & Bressan (1998) | 0.6–120 | 0.0004–0.05 | lifetimes (all), ejected mass + net yields (6–120), 9 elements |
| Marigo (2001) | 0.8–5 | 0.004–0.019 | ejected mass + net yields, C/N/O |
| Tumlinson, Shull & Venkatesan (2003) | 0.6–120 | 0 | lifetimes only |
| Heger & Woosley (2002) | 140–260 (He core) | 0 | PISN ejecta + 28 elements |

On the Ia side, `supernovaeTypeIaFixedYield` reads a **hard-coded** path (`Supernovae_Type_Ia_Yields.xml`, Nomoto et al. 1997 W7) with no `fileName` parameter and no metallicity axis, and the DTD is factored out into subclasses (`nagashima2005`, `powerLawDTD`, …). `numberCumulative(age,metallicity)` already carries `metallicity` in its interface — the power-law implementation just ignores it.

---

### Massive stars

#### Limongi & Chieffi (2018) — recommended, no code changes needed

The ORFEO database is alive (last updated March 2024) and hosts everything we need at
<https://orfeo.iaps.inaf.it/2018-modelli/main_2018.html>. I downloaded and parsed the tables to confirm:

- `yields/tab_{F,I,M,R}.tgz` → `tab_yieldsnet_ele_exp.dec` gives **net elemental yields in M☉** — exactly our convention, no conversion needed. 53 elements (H→Bi) × 9 masses (13, 15, 20, 25, 30, 40, 60, 80, 120) × 4 metallicities × 3 rotation velocities, as 12 stacked blocks in one plain-text file.
- `remas/remnant_masses.zip` gives remnant masses for all four explosion sets × 4 Z × 3 v_rot, so `ejectedMass = M_init − M_remnant` (a value of −1 flags a (P)PISN, i.e. no remnant).
- Metallicities recovered by summing the `initial` composition column: **Z = 1.345×10⁻², 3.236×10⁻³, 3.236×10⁻⁴, 3.236×10⁻⁵** for [Fe/H] = 0, −1, −2, −3. That extends us a factor ~12 below Portinari's floor of 4×10⁻⁴, which is the main gap James was pointing at.
- Lifetimes come from the paper's Table 5, which has a machine-readable version at IOP/ADS (it also carries pre-SN total mass, He and CO core masses).
- There is a newer extension to [Fe/H] = +0.3 (Limongi & Chieffi 2024, ApJS 271, 6) on the same site, which would extend us *above* solar too.

Two caveats, both real:

- **Explodability.** The four sets differ only in this. I confirmed the behaviour directly: in sets I and R everything above 25 M☉ collapses fully, so the net metal yield is ≈ 0 for M ≥ 30 (only the wind is ejected), whereas sets F and M force 0.07 M☉ of ⁵⁶Ni for every mass. These bracket the uncertainty, which is arguably a feature — I'd suggest shipping **set M** (mixing and fallback) as the headline table and **set R** as the alternative, and letting people swap via `XInclude`.
- **Mass range.** LC18 starts at 13 M☉, so the file has to be combined with an AGB/intermediate-mass table (Marigo, or better Karakas & Lugaro 2016 / Cristallo FRUITY) and a lifetime source. `XInclude` makes this trivial, which is exactly the flexibility the split-file refactor bought us.

Rotation is the interesting wrinkle: our table is indexed on (mass, metallicity) only, so v_rot has to be a *choice of file* rather than an interpolation axis. Three files (v0/v150/v300) is the natural approach; a rotation-distribution-averaged file would be a nice fourth.

#### Sukhbold et al. (2016) — worth having, but single-metallicity

Data at <https://wwwmpa.mpa-garching.mpg.de/ccsnarchive/data/SEWBJ_2015/>. Confirmed by download:

- `nucleosynthesis_yields.tar.gz` (41 MB) contains one `s<M>.yield_table` per model: 303 isotopes × two columns, `[ejecta]` and `[wind]`, as **gross** masses. Engines Z9.6, W18 and N20; W18 alone has 82 exploding + 105 imploding (BH-forming, wind-only) models, on a grid stepping 0.1 M☉ between 13 and 27 M☉. Vastly denser than LC18, and with a physically motivated explodability pattern, as James says.
- `M_ejected` falls straight out of summing the two columns (checked: the 20.1 M☉ W18 model gives 14.12 + 4.20 = 18.33 M☉).
- `progenitor_properties_presn.txt` gives pre-collapse, He-core and CO-core masses and compactness.

Costs relative to LC18:

- **Gross → net conversion required.** We need the KEPLER birth composition to form `net_i = gross_i − X_i,birth × M_ejected`.
- **Solar metallicity only, non-rotating.** This is the sharp edge: `fileMassYield` clamps the requested metallicity to the tabulated range, so a solar-only table would silently apply solar yields at *every* metallicity. (Note `lifetime` and `ejectedMass` are *not* clamped and would be extrapolated.) So S16 is only safe as an `XInclude` component alongside a metallicity-dependent set, or for deliberately controlled single-Z experiments. Worth stating loudly in the file's `<description>`.
- **No lifetimes**, so those would keep coming from Portinari/LC18.
- The grid density is ~190 models × ~50 elements. That is not a memory problem, but `Interpolate_2D_Irregular` is a Delaunay-type scheme over all tabulated points, so it deserves a timing check before we make it a default.

---

### Type Ia supernovae

All three of James' suggestions exist in convenient machine-readable form. Notably, VICE (James' own code, MIT-licensed) ships all of them in a uniform one-file-per-element layout, which is by far the easiest ingestion route; the originals are in the papers/HESMA if we'd rather go to source.

| Study | Models | Metallicity axis | Drop-in today? |
|---|---|---|---|
| Iwamoto et al. (1999) | 7 (W7, W70, WDD1–3, CDD1–2) | none | **Yes**, once `fileName` is a parameter |
| Seitenzahl et al. (2013) | 14 3-D delayed detonations (N1…N1600C) + N100 at Z/Z☉ = 0.01, 0.1, 0.5 | partial (N100 series) | Yes for a single model; the N100 series needs the new axis |
| Gronow et al. (2021a,b, 2022) | 44 = 11 (M_CO, M_He) × 4 metallicities (Z/Z☉ = 0.01, 0.1, 1, 3) | **yes** | No — needs format change |

One extra lead not in the thread: **Keegans et al. (2023, ApJS 268, 8)**, "Type Ia Supernova Nucleosynthesis: Metallicity-dependent Yields" — 39 models across **13** metallicities (²²Ne mass fractions 0 → 0.1) for three progenitor configurations, on Zenodo (10.5281/zenodo.8060323, CC-BY-4.0). Much finer metallicity sampling than Gronow, and it avoids the fixed-Fe-yield problem James flags. The catch is that Zenodo only holds LaTeX table sources, so it needs a parser.

Two small enabling changes are needed before any of this lands:

1. `supernovaeTypeIaFixedYield` should take a `fileName` input parameter instead of hard-coding `Supernovae_Type_Ia_Yields.xml`. That alone unlocks Iwamoto, Seitenzahl and single-model Gronow.
2. A metallicity-dependent variant of the yield file (a `<metallicity>` attribute per yield block, plus interpolation) to represent Gronow/Keegans. This is a new class rather than a change to `fixedYield`, since "fixed" is exactly the property being relaxed.

On rates and delay times:

- **Maoz & Mannucci (2012)** is already largely covered by `supernovaeTypeIaPowerLawDTD` (whose defaults come from Freundlich & Maoz 2021). Their three-bin measurement would slot in as a tabulated-DTD subclass if we want the observational constraint itself rather than a fit to it.
- **Johnson, Kochanek & Stanek (2023, MNRAS 526, 5911)** — the ~Z^(−1/2) rate enhancement at low metallicity. Because `numberCumulative(age,metallicity)` already receives the metallicity, this is a small multiplicative `(Z/Z☉)^β` factor: a new subclass of `supernovaeTypeIaMassIndependentDTD` with one extra parameter, ~30 lines.

### James' "parameterized yields" mode

This is the most interesting suggestion and the one that doesn't map cleanly onto the existing code. For Ia we're already there in spirit — `fixedYield` (yields) and the DTD subclasses (timing) are cleanly separated, so "user specifies a yield per element plus a DTD" is nearly free once (1) above is done.

For CCSNe there's an architectural mismatch worth being explicit about: Galacticus' `stellarAstrophysics` yields are **per-star** and IMF-integrated internally by `stellarPopulationStandard`, whereas the parameterization James describes is an **IMF-averaged population yield** y_X(Z). Forcing the latter through `stellarAstrophysics` would mean inventing a fake per-star distribution. The cleaner insertion point is a new `stellarPopulationProperties`/`stellarPopulation` implementation that supplies y_X(Z) and a recycled fraction directly, bypassing the IMF integral.

This lands on the same layer as #642 (generalizing the tracked "elements" to include other products of star formation, e.g. SN counts, via a new `stellarRemnants`-style class). The two are close to the same feature seen from opposite ends: a per-channel parameterized yield mode is naturally expressed as "for each enrichment channel, a yield per element plus a delay-time distribution", which needs exactly the per-channel event bookkeeping #642 proposes. If #642's generalized tracker lands first, the parameterized-yield mode largely reduces to registering CCSNe and SNe Ia as channels and attaching y_X(Z) to each. I'd suggest designing them together rather than separately.

---

### Proposed plan

**Phase 1 — enabling changes (small, unblocks everything else)**
1. Add a `fileName` input parameter to `supernovaeTypeIaFixedYield` (default unchanged, so no behaviour change).
2. Write converter scripts. There are none today — the existing XML files appear to have been built by hand — so these should live in the datasets repo (or `galacticusDevTools`) with the raw-table provenance recorded, so the tables can be regenerated.

**Phase 2 — Limongi & Chieffi (2018)** *(datasets-only; no Fortran)*
3. Convert ORFEO `tab_yieldsnet_ele_exp.dec` + `remnant_masses.txt` + Table 5 lifetimes into `stellarPropertiesLimongiChieffi2018_set{M,R}_v{0,150,300}.xml`.
4. Add `stellarPropertiesCompilationLimongiChieffi2018.xml` — an `XInclude` wrapper combining LC18 (M ≥ 13) with an AGB set and a lifetime source, mirroring the standard compilation.
5. Validate: compare IMF-integrated net metal yield and recycled fraction against the standard compilation over 10⁻⁵ ≤ Z ≤ 0.05; check the low-Z extension behaves sensibly rather than extrapolating.

**Phase 3 — Sukhbold et al. (2016)** *(datasets-only)*
6. Convert the W18 and N20 yield tables (incl. the wind-only imploding models), summing isotopes to elements and subtracting the birth composition to get net yields.
7. Ship *only* as an `XInclude` component with an explicit warning in `<description>` about the single metallicity; time a model run to confirm the dense grid doesn't hurt the irregular interpolation.

**Phase 4 — SN Ia yield tables** *(needs Phase 1.1)*
8. Add `Supernovae_Type_Ia_Yields_Iwamoto1999_{W7,W70,WDD1-3,CDD1-2}.xml` and `..._Seitenzahl2013_{N1…N1600C}.xml` in the existing format.
9. Add a metallicity-dependent Ia yield class + file format, and populate it from Gronow et al. and/or Keegans et al.

**Phase 5 — rates and delay times**
10. Add a metallicity-dependent DTD normalization, `(Z/Z☉)^β` with β ≈ −0.5, per Johnson, Kochanek & Stanek (2023).
11. Optionally add a tabulated-DTD class seeded with Maoz & Mannucci (2012).

**Phase 6 — parameterized yields** — track alongside #642 rather than as an independent effort; needs a design decision about where population-averaged yields enter (most likely a new `stellarPopulationProperties` implementation, not `stellarAstrophysics`).

Phases 2–4 are independent of each other and each is self-contained. Phase 2 is the highest value for the least work and is the natural starting point — and, because `fileName` and `XInclude` already exist, it lands entirely in the datasets repo.

#### Licensing / provenance

Worth settling before we import anything: ORFEO and the MPA archive publish data without an explicit licence (normal-citation practice), VICE's copies are MIT, and Keegans et al. is CC-BY-4.0. Each generated file should carry `<source>` and `<url>` tags — which the format already supports and the existing files already use — and the converter scripts should record the exact upstream file and download date.

**Comment by abensonca** [comment]
The Sukhbold et al. (2016) models are now working, and the diagnosis is worth recording because the failure mode is general and silent.

### What was wrong

Not the grid density, and not a bug in the interpolation as such. `stellarAstrophysicsFile` interpolates irregularly in the (mass, metallicity) plane, and the scheme extrapolates badly when the requested metallicity falls outside the range spanned by the tabulated points **at the relevant masses**. For a table computed at a single metallicity that is almost always the case, because there is nothing above or below it to interpolate between.

That one cause accounts for every symptom, which is why thinning the grid did not help:

| configuration | points near 20 M☉ | request at Z ≈ 0.019 | result |
|---|---|---|---|
| Sukhbold + Heger & Woosley (2002) | Z = 0 and 0.0149 | above both | 11.4× the expected metals |
| Sukhbold alone | Z = 0.0149 only | outside a one-dimensional set | zero |
| Sukhbold thinned to 20 models | Z = 0.0149 only | non-finite | hangs inside the IMF integration |
| Sukhbold bracketed | Z = 10⁻⁴ and 10⁻¹ | inside | 0.61× — correct |

### The fix

Each model is now written twice, at two bracketing metallicities with identical yields. That is not a workaround dressed up: it is the explicit statement that these yields do not depend on metallicity, which is exactly what a single-metallicity table means when it is used across a range of metallicities. Every request then becomes an interpolation.

The result checks out independently. The compilation returns 0.61× the metals of the standard one; solving for the contribution common to both compilations gives 0.0054 for the AGB plus Type Ia share, about 16% of the total, which is what one would expect. Stellar masses agree to 0.04%. `testSuite/test-stellar-properties-Sukhbold2016.py` pins this, with a band that excludes all three failure modes above.

The compilation file previously withheld is now shipped, with the standing caveat that the yields remain metallicity-independent — a limitation of the models, not of the encoding.

### The general trap

**A yield table whose metallicity coverage does not span the range a model requests fails silently.** No error, no warning: just wrong numbers, or zero, or an integrator that subdivides without bound until it exhausts memory. I hit the last of these for real — two runaway processes sat on this machine for about sixteen hours and contributed to an out-of-memory kill of an unrelated build.

Anything added later that is computed at one metallicity, or over a range narrower than models will reach, will hit this. A guard in `stellarAstrophysicsFile` that checked, at read time, whether each property's metallicity coverage brackets the plausible range and reported it plainly would turn a silent corruption into an error message. Recording that here rather than opening a separate issue.

### On performance, with a correction

I expected this to be largely a performance problem and it was not, so the earlier note about timing needs qualifying.

Each interpolated yield now has its own interpolation workspace, where previously one was shared between the total metal yield and every element. Since each is tabulated at the same points but with different values, the shared workspace was invalidated on every call and the partial derivatives re-estimated each time, at a cost of O(N n²). A `dataStatic` argument now lets a caller whose data are read once say so.

The gain is real but smaller than that analysis implied: the Sukhbold compilation falls from 318 s to 221 s, while the standard and Limongi & Chieffi compilations are unchanged within the scatter from other jobs on the same machine. So derivative re-estimation was a significant cost but not the dominant one, and whatever limits the large tables is elsewhere. I have not chased it further. Results are unchanged throughout.

Two smaller things fixed alongside: `Interpolate_2D_Irregular` now rebuilds its workspace when the number of data points differs from the number it was built with, rather than leaving the triangulation describing a different point set; and `fileMassYield` no longer indexes its range arrays with the element index before checking it, which is −1 for an element the compilation supplies no yields for.
